enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,281 @@
+a container-runtime component (CRI-O in OpenShift case).
+
+### API
+The PerformanceProfile extended with a new field under `spec.cpu`:
+
+`shared` - Shared define the set of CPUs that can be allocated to guaranteed pods/containers that still require
+non-exclusive ones.
+The `shared` field follows the same rules as the other fields, meaning it cannot overlap the other
+cpu fields, contains only integral numbers, and can be specified as [cpuset](https://github.com/openshift/cluster-node-tuning-operator/blob/master/docs/performanceprofile/performance_profile.md#cpuset) conventions.
+The field is optional and None by default.
+
+```yaml
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: example-performanceprofile
+spec:
+  cpu:
+    reserved: "0-1"
+    shared: "2-3"
+    isolated: "4-8"
+  hugepages:
+    defaultHugepagesSize: "1G"
+    pages:
+      - size: "1G"
+        count: 2
+        node: 0
+  realTimeKernel:
+    enabled: true
+  nodeSelector:
+    node-role.kubernetes.io/performance: "test"
+```
+
+Specify `shared` cpus activates the feature, signals NTO to update the `reserved-cpus` in Kubelet config, 
+and deploys the mixed-cpu-node-plugin.
+The logic of updating reserved-cpus pool is NTO's responsibility and would be implemented as part of this feature.
+
+Remove the `shared` reverts the changes from the `reserve-cpus`
+and deletes the mixed-cpu-node-plugin from the cluster.
+
+### Workflow Description
+
+The premise is that OCP cluster is running, and there's an active performance-profile that already tuned the system.
+
+1. The cluster administrator wants to support both exclusive and shared cpus for workloads running on the cluster.
+2. The cluster administrator specifies the shared CPU ids under the PerformanceProfile's `spec.cpu.shared` field.
+3. The cluster administrator waits for MCO to kick in and update the nodes.
+4. The application administrator wants to deploy its DPDK application as a Guaranteed pod with shared CPUs.
+5. The application administrator specifies a request for `openshift.io/shared-cpus: 1` under the pod's `spec.containers[].resources.requests`.
+6. The application administrator is waiting for the DPDK pod to be `Running`.
+7. The DPDK's app user/developer wants to run a light-weight task (threads) on shared cpus.
+8. The DPDK's app user/developer should pin the light-weight threads to CPUs that have shown in the `OPENSHIFT_SHARED_CPUS` environment variable of the container's process.
+
+### API Extensions
+N/A
+
+### Risks and Mitigations
+
+* With this solution, the platform's housekeeping processes also runs on the shared CPUs.
+but the intent is to allocate those cpus to workload’s light-weight tasks,
+so having some latency is bearable.
+A way of mitigating that is checks whether it's possible to use workload partitioning
+to ensure the platform housekeeping processes don't run on the shared cpus.
+(This statement needs to be verified)
```

Comment:
In any case, I think this proposal needs to discuss the interactions with workload partitioning, which currently places management workloads on the reserved cpus. Will workload partitioning use the larger set of reserved cpus that now includes the "shared" cpus? What happens if the new shared cpu set is created at installation time vs. on an already running system? I think there are implications for changing the reserved cpuset after workload partitioning is configured.

Reply:
Yea, this is still under investigation. 
>Will workload partitioning use the larger set of reserved cpus that now includes the "shared" cpus? 

@bartwensley @browsell  If we would make sure (as part of this feature) that housekeeping processes would run only on a subset of the reserved that doesn't include the shared cpus, would that be a more satisfying  approach for this solution?

Reply:
This is the [file](https://github.com/redhat-ztp/ztp-cluster-deploy/blob/master/ai-deploy-cluster-singlenode/roles/enable-workload-partitioning/templates/kubelet.conf) that contains management workloads when  SNO spin ups with WP enabled.
According to my understanding changing the reserved pool later on, won't affect this file, because it's a static configuration, hence houskeeping processes won't be expanded to the shared-cpus.
I asked for SNO cluster for testing that theory but  please keep me honest here if I presumed something wrong.

Reply:
First of all, workload partitioning is not just SNO. The cpuset in the crio dropin for management workload partitioning is generated based on the number of reserved cpus in the performance profile,  https://github.com/openshift/cluster-node-tuning-operator/blob/master/pkg/performanceprofile/controller/performanceprofile/components/machineconfig/machineconfig.go#L530.

System reserved with this proposal no longer maps to kubelet's view of system reserved. As long as the NTO continues to generate the crio drop in based on the system reserved defined in the performance profile then there should not be an issue.

Reply:
Thank you Brent for your comment.
>As long as the NTO continues to generate the crio drop in based on the system reserved defined in the performance profile then there should not be an issue.

With this proposal, in kubelet configuration file the `reservedSystemCpus` composed of preformanceProfile's `.spec.cpu.reserved` + `.spec.cpu.shared`.
But as you said the generated crio drop in, would be composed only of  preformanceProfile's `.spec.cpu.reserved`

Do we have an agree on that?

Reply:
this comment thread is useful and the key point about crio.conf config separating which cpus are reserved for management component use from the other cpus now available in the shared pool is clarifying.